scalardl/src/jepsen/scalardl.clj

```diff
@@ -0,0 +1,159 @@
+(ns jepsen.scalardl
+  (:require [clojure.tools.logging :refer [debug info warn]]
+            [jepsen
+             [cli :as cli]
+             [client :as client]
+             [checker :as checker]
+             [control :as c]
+             [db :as db]
+             [generator :as gen]
+             [independent :as independent]
+             [nemesis :as nemesis]
+             [tests :as tests]]
+            [jepsen.control.util :as cu]
+            [jepsen.scalardl.support :as s]
+            [jepsen.os.debian :as debian]
+            [knossos.model :as model]
+            [knossos.op])
+  (:import (com.scalar.client.service ClientService)
+           (java.util Optional)))
+
+(defn parse-long
+  [s]
+  (when s (Long/parseLong s)))
+
+(def dir "/opt/scalardl")
+(def binary "bin/scalar-ledger")
+(def logfile (str dir "/scalardl.log"))
+(def pidfile (str dir "/scalardl.pid"))
+(def ledger-dir "/jepsen/scalardl/ledger")
+
+(defn r [_ _] {:type :invoke, :f :read, :value nil})
+(defn w [_ _] {:type :invoke, :f :write, :value (rand-int 5)})
+
+(defrecord Client [conn]
+  client/Client
+  (setup! [this test]
+    (info "registering certificates and contracts")
+    (.registerCertificate conn)
+    (.registerContract
+      conn
+      "read"
+      "com.scalar.jepsen.scalardl.Read"
+      (str ledger-dir "/Read.class")
+      (Optional/empty))
+    (.registerContract
+      conn
+      "write"
+      "com.scalar.jepsen.scalardl.Write"
+      (str ledger-dir "/Write.class")
+      (Optional/empty)))
+
+  (open! [this test node]
+    (let [clientService (.getInstance (s/create-injector node) ClientService)]
+      (assoc this :conn clientService)))
+
+  (close! [this test]
+    (.close conn))
+
+  (invoke! [_ test op]
+    (let [[k v] (:value op)]
+      (case (:f op)
+        :read (let [argument (s/create-argument k)
+                    value (s/response->int (.executeContract conn "read" argument))]
+                (assoc op :type :ok, :value (independent/tuple k value)))
```

Comment:
Doesn't `read` fail?

Reply:
Yes, I should be checking the `ContractExecutionResponse` for 200 and anything else should fail. For some reason I thought it would throw an exception. Thank you. Good catch.